packages/async-flow/src/async-flow.js

```diff
@@ -0,0 +1,453 @@
+            //
+            // `bijection.hasGuest(guestResultP)` can be false in a delayed
+            // guest - to - host settling from a previous run.
+            // In that case, the bijection was reset and all guest caps
+            // created in the previous run were unregistered,
+            // including `guestResultP`.
+            void E.when(
+              guestResultP,
+              gFulfillment => {
+                if (bijection.hasGuest(guestResultP)) {
+                  outcomeKit.resolver.resolve(
+                    membrane.guestToHost(gFulfillment),
+                  );
+                  admin.complete();
+                }
+              },
+              guestReason => {
+                // The `guestResultP` might be a failure thrown by `panic`
+                // indicating a failure to replay. In that case, we must not
+                // settle the outcomeVow, since the outcome vow only represents
+                // the settled result of the async function itself.
+                // Fortunately, `panic` resets the bijection, again resulting
+                // in the `guestResultP` being absent from the bijection,
+                // so this leave the outcome vow unsettled, as it must.
+                if (bijection.hasGuest(guestResultP)) {
+                  outcomeKit.resolver.reject(membrane.guestToHost(guestReason));
+                  admin.complete();
+                }
+              },
+            );
+          },
+          wake() {
+            const { facets } = this;
+            const { flow } = facets;
+            const eph = tmp.for(flow);
+
+            const flowState = flow.getFlowState();
+            if (flowState === 'Done' || flowState === 'Failed') {
+              return;
+            }
+            if (eph.membrane) {
+              eph.membrane.wake();
+            } else {
+              flow.restart();
+            }
+          },
+          getOutcome() {
+            const { state } = this;
+            const { outcomeKit } = state;
+            return outcomeKit.vow;
+          },
+          dump() {
+            const { state } = this;
+            const { log } = state;
+
+            return log.dump();
+          },
+          getOptFatalProblem() {
+            const { facets } = this;
+            const { flow } = facets;
+
+            return failures.has(flow) ? failures.get(flow) : undefined;
+          },
+        },
+        admin: {
+          reset() {
+            const { state, facets } = this;
+            const { bijection, log } = state;
+            const { flow } = facets;
+            const eph = tmp.for(flow);
+
+            if (failures.has(flow)) {
+              failures.delete(flow);
+            }
+            if (eph.membrane) {
+              eph.membrane.stop();
+            }
+            tmp.resetFor(flow);
+            log.reset();
+            bijection.reset();
+
+            state.isDone = false;
```

Comment:
Why are we changing `isDone`? I know `restart()` is the only place where this is called from a public path, and it checks that we're not done first, but maybe we should move that check here and never set `isDone` to false (except at init)?

Reply:
Changed to an assertion. Done.